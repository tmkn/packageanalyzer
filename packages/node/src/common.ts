import * as path from "path";
import * as fs from "fs";
import { fileURLToPath } from "url";

import dayjs from "dayjs";

import type { DependencyTypes } from "../../shared/src/reports/Validation.js";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const __dirname = path.dirname(fileURLToPath(import.meta.url));
        // adjust when moving this function to another folder
        const file = path.join(__dirname, "..", "..", "..", "package.json");
        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}
