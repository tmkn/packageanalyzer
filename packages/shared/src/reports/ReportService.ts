import chalk from "chalk";

import { type IPackage } from "../package/package.js";
import { Formatter, type IFormatter } from "../utils/formatter.js";
import { Visitor } from "../visitors/visitor.js";
import { type GenericReport, type IReportConfig, isReportConfigArray } from "./Report.js";
import type { IHost } from "../host/IHost.js";

export type ReportServiceConfig = {
    reports: GenericReport[];
};

export class ReportService {
    private readonly _stdoutFormatter: IFormatter;
    private readonly _stderrFormatter: IFormatter;

    constructor(
        private readonly _config: ReportServiceConfig,
        private readonly _host: IHost
    ) {
        this._stdoutFormatter = new Formatter(this._host.getStdoutWriter());
        this._stderrFormatter = new Formatter(this._host.getStderrWriter());
    }

    async process(): Promise<number | void> {
        let exitCode: number = 0;

        try {
            exitCode = await this._report(this._config);
        } catch (e: any) {
            this._stderrFormatter.writeLine(e?.toString());
            console.error(e?.toString());

            exitCode = 1;
        } finally {
            await Promise.all([
                this._host.getStdoutWriter().flush(),
                this._host.getStderrWriter().flush()
            ]);
        }

        return exitCode;
    }

    private async _report({ reports }: ReportServiceConfig): Promise<number> {
        for (const report of reports) {
            const packages: IPackage[] = await this._getPackages(report);

            if (reports.length > 1)
                this._stdoutFormatter.writeLine(chalk.underline.bgBlue(`Report: ${report.name}`));

            await report.report(packages, {
                stdoutFormatter: this._stdoutFormatter,
                stderrFormatter: this._stderrFormatter
            });
            this._stdoutFormatter.writeLine(``);
        }

        return Math.max(...reports.map(report => report.exitCode));
    }

    private async _getPackage(entry: IReportConfig, report: GenericReport): Promise<IPackage> {
        const visitor = new Visitor(
            entry.pkg,
            report.provider ?? this._host.getDefaultProvider(),
            this._host.getLogger(),
            entry.attachments ?? {},
            entry.depth
        );

        return visitor.visit(entry.type);
    }

    private async _getPackages(report: GenericReport): Promise<IPackage[]> {
        const entries: Array<IReportConfig> = isReportConfigArray(report.configs)
            ? report.configs
            : [report.configs];
        const packageArgs: IPackage[] = [];

        for (const entry of entries) {
            packageArgs.push(await this._getPackage(entry, report));
        }

        return packageArgs;
    }
}
