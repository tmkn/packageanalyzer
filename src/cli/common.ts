import * as path from "path";
import * as fs from "fs";

import * as dayjs from "dayjs";
import * as chalk from "chalk";

import { PackageAnalytics, VersionSummary, GroupedLicenseSummary } from "../analyzers/package";
import { DependencyTypes } from "../visitors/visitor";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const file = path.join(__dirname, "./../../package.json");

        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch (e) {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

export function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(chalk.bold(`Statistics for ${pa.fullName}\n`));

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa, padding, paddingLeft);
    printLicenseInfo(pa.licensesByGroup, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest package:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
        console.log(`${`Newest package path:`.padEnd(padding)}${newest.pathString}`);
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest package:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
        console.log(`${`Oldest package path:`.padEnd(padding)}${oldest.pathString}`);
    }
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(pa: PackageAnalytics, padding: number, paddingLeft: number): void {
    const { loops, loopPathMap, distinctLoopCount } = pa;

    console.log(`${`Loops:`.padEnd(padding)}${loops.length} (${distinctLoopCount} distinct)`);

    if (distinctLoopCount > 0) {
        const [first] = loops.map(l => l.pathString).sort();

        console.log(`    affected Packages: [${[...loopPathMap.keys()].join(", ")}]`);
        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(groupedLicenses: GroupedLicenseSummary, paddingLeft: number): void {
    console.log(`Licenses:`);
    for (const { license, names } of groupedLicenses) {
        const threshold = 4;
        const samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${
                    names.length - threshold
                } more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }
}
