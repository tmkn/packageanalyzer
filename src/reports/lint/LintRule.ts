import { type ZodTypeAny, z } from "zod";

import type { IPackage } from "../../package/package.js";
import type { AttachmentData, Attachments } from "../../attachments/Attachments.js";

const LintTypes = z.union([z.literal("error"), z.literal("warning")]);

export type ILintTypes = z.infer<typeof LintTypes>;

interface ISharedLintCheckParams {
    name: string;
    checkParams?: () => ZodTypeAny;
}

export interface IBaseLintCheck<T> extends ISharedLintCheckParams {
    check: (pkg: IPackage, params: T) => string | string[] | void;
}

export interface ILintCheckWithAttachments<T, A extends Attachments>
    extends ISharedLintCheckParams {
    check: (pkg: IPackage<AttachmentData<A>>, params: T) => string | string[] | void;
    attachments: A;
}

export type ILintCheck<T = undefined, A extends Attachments | undefined = undefined> = A extends {}
    ? ILintCheckWithAttachments<T, A>
    : IBaseLintCheck<T>;

type IBaseLintTuple<P> = P extends {}
    ? [ILintTypes, IBaseLintCheck<P>, P]
    : [ILintTypes, IBaseLintCheck<P>];

type ILintTupleWithAttachmentsTuple<P, A extends Attachments> = P extends {}
    ? [ILintTypes, ILintCheckWithAttachments<P, A>, P]
    : [ILintTypes, ILintCheckWithAttachments<P, A>];

// infer the correct LintCheck type
// it somehow does what it's supposed to do lol
export type LintRuleTuple<C extends ILintCheck<any, any> = ILintCheck> =
    C extends ILintCheck<infer P, infer A extends Attachments>
        ? ILintTupleWithAttachmentsTuple<P, A>
        : C extends ILintCheck<infer P>
          ? IBaseLintTuple<P>
          : never;

export function hasAttachments<P>(
    check: IBaseLintCheck<P> | ILintCheckWithAttachments<P, any>
): check is ILintCheckWithAttachments<P, any> {
    return "attachments" in check;
}

export const ZodLintRule = z.custom<LintRuleTuple<ILintCheck<any>>>(data => {
    if (Array.isArray(data)) {
        const [type, check] = data;

        const validType = LintTypes.safeParse(type).success;
        const validCheck =
            typeof check === "object" &&
            typeof check.name === "string" &&
            typeof check.check === "function";

        return validType && validCheck;
    }

    return false;
});

// todo unify createRule and createRuleWithAttachment for better developer experience
export function createRule<P = undefined>(...args: IBaseLintTuple<P>): IBaseLintTuple<P> {
    return [...args];
}

export function createRuleWithAttachment<P = undefined, A extends Attachments = {}>(
    ...args: ILintTupleWithAttachmentsTuple<P, A>
): ILintTupleWithAttachmentsTuple<P, A> {
    return [...args];
}

export const LintFile = z.object({
    rules: z.array(ZodLintRule)
});

export type ILintFile = z.infer<typeof LintFile>;
