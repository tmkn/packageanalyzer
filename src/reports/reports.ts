import { IDecorator } from "../extensions/decorators/Decorator";
import { Package } from "../package/package";
import { IPackageVersionProvider } from "../providers/folder";
import { IFormatter } from "../utils/formatter";
import { PackageVersion } from "../visitors/visitor";

export interface IReport<T extends {}> {
    readonly name: string;
    readonly params: T;
    readonly pkg: PackageVersion;

    decorators?: IDecorator<any>[];
    provider?: IPackageVersionProvider;

    report(pkg: Package, formatter: IFormatter): Promise<void>;
}

interface ITestReportParams {}
export class TestReport implements IReport<ITestReportParams> {
    name = `Test Report`;
    pkg: PackageVersion = [``, ``];

    constructor(readonly params: ITestReportParams) {}

    async report(pkg: Package, formatter: IFormatter): Promise<void> {}
}

export class ReportService {
    async process(reports: IReport<any>[]): Promise<void> {
        try {
            for (const report of reports) {
            }
        } catch (e) {}
    }
}
