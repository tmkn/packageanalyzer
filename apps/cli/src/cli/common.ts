import { Command } from "clipanion";

import { AbstractReport } from "../../../../packages/shared/src/reports/Report.js";
import { ReportService } from "../../../../packages/shared/src/reports/ReportService.js";
import { Formatter, type IFormatter } from "../../../../packages/shared/src/utils/formatter.js";
import { NodeHost } from "../../../../packages/node/src/host/NodeHost.js";

export abstract class CliCommand<
    T extends AbstractReport<any> | AbstractReport<any>[]
> extends Command {
    abstract getReports(): T | Promise<T>;

    exitCode = 0;

    beforeProcess: ((report: T) => void) | undefined = undefined;

    async execute(): Promise<number | void> {
        const host = new NodeHost(this.context.stdout, this.context.stderr);

        try {
            const reports = await this.getReports();
            const reportService = new ReportService(
                {
                    reports: Array.isArray(reports) ? reports : [reports]
                },
                host
            );

            this.beforeProcess?.(reports);
            this.exitCode = (await reportService.process()) ?? 0;
        } catch (e: unknown) {
            const stderrFormatter: IFormatter = new Formatter(host.getStderrWriter());

            if (e) stderrFormatter.writeLine(e?.toString());
            this.exitCode = 1;
        } finally {
            await Promise.all([host.getStdoutWriter().flush(), host.getStderrWriter().flush()]);
        }

        return this.exitCode;
    }
}
