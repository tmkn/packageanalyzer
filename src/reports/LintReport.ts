import * as path from "path";
import * as chalk from "chalk";
import { z } from "zod";

import { Package } from "../package/package";
import { AbstractReport, IReportContext } from "./Report";
import { PathUtilities } from "../extensions/utilities/PathUtilities";
import { ZodLintRule } from "./lint/LintRule";
import { ILintResult, LintResultFormatter } from "./lint/LintResultFormatter";
import { PackageVersion, getPackageVersionfromString } from "../visitors/visitor";
import { getPackageVersionFromPath } from "../visitors/util.node";
import { FileSystemPackageProvider } from "../providers/folder";

const LintFile = z.object({
    rules: z.array(ZodLintRule)
});

type ILintFile = z.infer<typeof LintFile>;

const BaseLintParams = z.object({
    lintFile: z.string(),
    depth: z.number()
});

const FolderLintParams = BaseLintParams.merge(z.object({ folder: z.string() }));
const PackageLintParams = BaseLintParams.merge(z.object({ package: z.string() }));

export const LintReportParams = z.union([FolderLintParams, PackageLintParams]);

export type ILintParams = z.infer<typeof LintReportParams>;

export class LintReport extends AbstractReport<ILintParams> {
    name = `Lint Report`;
    pkg: PackageVersion;
    override depth: number | undefined;

    rules: ILintFile;

    constructor(params: ILintParams) {
        super(params);
        this.depth = params.depth;

        if (this._isPackageLintParams(params)) {
            this.pkg = getPackageVersionfromString(params.package);
        } else {
            this.pkg = getPackageVersionFromPath(params.folder);
            this.provider = new FileSystemPackageProvider(params.folder);
        }

        const lintFile = this._loadLintFile(params.lintFile);

        if (this._isLintFile(lintFile)) {
            this.rules = lintFile;
        } else {
            throw new Error(`Invalid lint file format: ${params.lintFile}`);
        }
    }

    private _loadLintFile(lintFile: string): unknown {
        let lintFileData: unknown;

        try {
            const lintFilePath = path.join(process.cwd(), lintFile);

            const data = require(lintFilePath);
            lintFileData = data;
        } catch (e) {
            throw new Error(`Couldn't find lint file: ${lintFile}`);
        }

        return lintFileData;
    }

    private _isPackageLintParams(data: unknown): data is z.infer<typeof PackageLintParams> {
        return PackageLintParams.safeParse(data).success;
    }

    private _isLintFile(data: unknown): data is z.infer<typeof LintFile> {
        return LintFile.safeParse(data).success;
    }

    async report(context: IReportContext, pkg: Package): Promise<void> {
        const resultFormatter = new LintResultFormatter(context.stdoutFormatter);
        const lintResults: ILintResult[] = [];

        context.stdoutFormatter.writeLine(`PackageLint: ${pkg.fullName}`);

        pkg.visit(dep => {
            for (const [type, rule, params] of this.rules.rules) {
                const checkResult = rule.check(dep, params);

                if (typeof checkResult === `string` || Array.isArray(checkResult)) {
                    if (type === `error`) {
                        process.exitCode = 1;
                    }

                    //beautify path
                    const path: string[] = new PathUtilities(dep).path
                        .map(([name, version]) => `${name}@${version}`)
                        .map(name => chalk.cyan(name));

                    const pathString: string = path.join(chalk.white(` â†’ `));

                    const messages = Array.isArray(checkResult) ? checkResult : [checkResult];

                    for (const message of messages) {
                        lintResults.push({
                            type,
                            name: rule.name,
                            message,
                            path: pathString,
                            pkg: dep,
                            rootPkg: pkg.fullName
                        });
                    }
                }
            }
        }, true);

        resultFormatter.format(lintResults);
    }

    override validate(): z.ZodTypeAny {
        return LintReportParams;
    }
}
