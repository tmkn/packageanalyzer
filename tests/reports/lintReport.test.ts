import path from "path";

import { z } from "zod";
import { ZodLintRule, createRule } from "../../src/reports/lint/LintRule";
import { ILintFile, LintReport } from "../../src/reports/LintReport";
import { ReportService } from "../../src/reports/ReportService";
import { createMockContext } from "../common";
import { IMockPackageJson, MockProvider } from "../mocks";
import exp from "constants";

describe(`LintReport Test`, () => {
    const medalloPkg: IMockPackageJson = {
        name: `medallo`,
        version: `1.0.0`,
        dependencies: [
            { name: `foo`, version: `1.0.0` },
            { name: `bar`, version: `1.0.0` }
        ]
    };
    const provider = new MockProvider([medalloPkg]);

    beforeEach(() => {
        jest.resetModules();
    });

    test(`loads and executes lint file with absolute path`, async () => {
        const mockFn = jest.fn();
        jest.doMock(
            `/getsMockedAnyway.js`,
            () => ({
                rules: [createRule(`error`, { name: `test-rule`, check: mockFn })]
            }),
            {
                virtual: true
            }
        );

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: Infinity
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        await reportService.process();

        expect(mockFn).toHaveBeenCalledTimes(3);
    });

    test(`loads and executes lint file with relative path`, async () => {
        const mockFn = jest.fn();
        jest.doMock(
            path.join(process.cwd(), `./getsMockedAnyway.js`),
            () => ({
                rules: [createRule(`error`, { name: `test-rule`, check: mockFn })]
            }),
            {
                virtual: true
            }
        );

        const lintReport = new LintReport({
            lintFile: `./getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: Infinity
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        await reportService.process();

        expect(mockFn).toHaveBeenCalledTimes(3);
    });

    test(`errors on invalid lint file`, async () => {
        jest.doMock(`/getsMockedAnyway.js`, () => ({}), {
            virtual: true
        });

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: Infinity
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        expect.assertions(2);
        const exitCode = await reportService.process();

        expect(stderr.lines).toMatchSnapshot();
        expect(exitCode).toBe(1);
    });

    test(`calls lint function only on root`, async () => {
        const mockFn = jest.fn();
        jest.doMock(
            `/getsMockedAnyway.js`,
            () => ({
                rules: [createRule(`error`, { name: `test-rule`, check: mockFn })]
            }),
            {
                virtual: true
            }
        );

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: 0
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        await reportService.process();

        expect(mockFn).toHaveBeenCalledTimes(1);
    });

    test(`uses custom validation from lint check`, async () => {
        const validateFn = jest.fn();
        jest.doMock(
            `/getsMockedAnyway.js`,
            () => ({
                rules: [
                    createRule(
                        `error`,
                        {
                            name: `test-rule`,
                            check: () => {},
                            checkParams: () => z.custom(validateFn)
                        },
                        { name: `medallo` }
                    )
                ]
            }),
            {
                virtual: true
            }
        );

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: 0
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        await reportService.process();

        expect(validateFn).toHaveBeenCalledTimes(1);
    });

    test(`correctly throws on failed custom validation`, async () => {
        jest.doMock(
            `/getsMockedAnyway.js`,
            () => ({
                rules: [
                    createRule(
                        `error`,
                        {
                            name: `test-rule`,
                            check: () => {},
                            checkParams: () => z.custom(() => false)
                        },
                        "some param"
                    )
                ]
            }),
            {
                virtual: true
            }
        );

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: 0
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        expect.assertions(1);
        await reportService.process();

        expect(stdout.lines).toMatchSnapshot();
    });

    test(`calls lint function with custom params`, async () => {
        jest.doMock(
            `/getsMockedAnyway.js`,
            () => ({
                rules: [
                    createRule(
                        `error`,
                        {
                            name: `test-rule`,
                            check: (_, params) => {
                                expect(params.foo).toBe(`bar`);
                            }
                        },
                        { foo: `bar` }
                    )
                ]
            }),
            {
                virtual: true
            }
        );

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: Infinity
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        expect.assertions(3);
        await reportService.process();
    });

    test(`sets exitCode to 1 if at least 1 error is reported`, async () => {
        jest.doMock(
            `/getsMockedAnyway.js`,
            () => ({
                rules: [
                    createRule(`warning`, { name: `test-rule-warning`, check: () => `` }, {}),
                    createRule(`error`, { name: `test-rule-error`, check: () => `` }, {})
                ]
            }),
            {
                virtual: true
            }
        );

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: Infinity
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        const exitCode = await reportService.process();

        expect(exitCode).toBe(1);
    });

    test(`sets exitCode to 0 if only warnings are reported`, async () => {
        jest.doMock(
            `/getsMockedAnyway.js`,
            () => ({
                rules: [createRule(`warning`, { name: `test-rule-warning`, check: () => `` }, {})]
            }),
            {
                virtual: true
            }
        );

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: Infinity
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        const exitCode = await reportService.process();

        expect(exitCode).toBe(0);
    });
});

describe(`"internal-error" Lint Test`, () => {
    const medalloPkg: IMockPackageJson = {
        name: `medallo`,
        version: `1.0.0`,
        dependencies: [
            { name: `foo`, version: `1.0.0` },
            { name: `bar`, version: `1.0.0` }
        ]
    };
    const provider = new MockProvider([medalloPkg]);

    const lintFile: ILintFile = {
        rules: [
            createRule(`warning`, { name: `report-all`, check: () => `` }, {}),
            createRule(
                `warning`,
                {
                    name: `will-throw`,
                    check: () => {
                        throw new Error(`whoops`);
                    }
                },
                {}
            )
        ]
    };

    beforeEach(() => {
        jest.resetModules();
    });

    test(`sets exitCode to 1 if there are internal errors`, async () => {
        jest.doMock(`/getsMockedAnyway.js`, () => lintFile, {
            virtual: true
        });

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: Infinity
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        const exitCode = await reportService.process();

        expect(exitCode).toBe(1);
    });

    test(`prints out internal errors & exit message`, async () => {
        jest.doMock(`/getsMockedAnyway.js`, () => lintFile, {
            virtual: true
        });

        const lintReport = new LintReport({
            lintFile: `/getsMockedAnyway.js`,
            package: `medallo@1.0.0`,
            depth: Infinity
        });
        lintReport.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                reports: [lintReport]
            },
            stdout,
            stderr
        );

        await reportService.process();

        expect(stdout.lines).toMatchSnapshot();
        expect(stderr.lines).toMatchSnapshot();
    });
});

describe(`Lint Rule Validation`, () => {
    test(`parses error rule`, () => {
        const rule = [
            `error`,
            {
                name: `test-rule`,
                check: () => {}
            }
        ];

        expect(ZodLintRule.safeParse(rule).success).toBe(true);
    });

    test(`parses warning rule`, () => {
        const rule = [
            `warning`,
            {
                name: `test-rule`,
                check: () => {}
            }
        ];

        expect(ZodLintRule.safeParse(rule).success).toBe(true);
    });

    test(`fails on invalid type`, () => {
        const rule = [
            `abc`,
            {
                name: `test-rule`,
                check: () => {}
            }
        ];

        expect(ZodLintRule.safeParse(rule).success).toBe(false);
    });

    test(`fails on missing name`, () => {
        const rule = [
            `warning`,
            {
                check: () => {}
            }
        ];

        expect(ZodLintRule.safeParse(rule).success).toBe(false);
    });

    test(`fails on missing check callback`, () => {
        const rule = [
            `warning`,
            {
                name: `test-rule`
            }
        ];

        expect(ZodLintRule.safeParse(rule).success).toBe(false);
    });

    test(`fails on missing lint rule`, () => {
        const rule = [`warning`];

        expect(ZodLintRule.safeParse(rule).success).toBe(false);
    });

    test(`fails on empty lint rule`, () => {
        const rule = [`warning`, {}];

        expect(ZodLintRule.safeParse(rule).success).toBe(false);
    });

    test(`fails on invalid rule defs`, () => {
        expect(ZodLintRule.safeParse({}).success).toBe(false);
        expect(ZodLintRule.safeParse([]).success).toBe(false);
        expect(ZodLintRule.safeParse(``).success).toBe(false);
        expect(ZodLintRule.safeParse(23).success).toBe(false);
        expect(ZodLintRule.safeParse(true).success).toBe(false);
    });
});
