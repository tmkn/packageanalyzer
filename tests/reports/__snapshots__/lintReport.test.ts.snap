// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"internal-error" Lint Test prints out internal errors & exit message 1`] = `
[
  "PackageLint: medallo@1.0.0",
  "",
  "medallo@1.0.0",
  "  [warning][medallo@1.0.0][report-all]: ",
  "  [internal-error][medallo@1.0.0][will-throw]: whoops",
  "",
  "medallo@1.0.0 → foo@1.0.0",
  "  [warning][foo@1.0.0][report-all]: ",
  "  [internal-error][foo@1.0.0][will-throw]: whoops",
  "",
  "medallo@1.0.0 → bar@1.0.0",
  "  [warning][bar@1.0.0][report-all]: ",
  "  [internal-error][bar@1.0.0][will-throw]: whoops",
  "",
  "Found 3 warning(s) and 0 error(s)",
  "Terminated with 3 internal error(s), please check lint output",
  "",
  "",
]
`;

exports[`"internal-error" Lint Test prints out internal errors & exit message 2`] = `[]`;
